src/Uno.UI.Runtime.Skia.MacOS/MacOSNativeElementHostingExtension.cs

```diff
@@ -0,0 +1,157 @@
+#nullable enable
+
+using Windows.Foundation;
+using Windows.UI.Core;
+using Microsoft.UI.Xaml.Controls;
+
+using Uno.Foundation.Extensibility;
+using Uno.Foundation.Logging;
+
+namespace Uno.UI.Runtime.Skia.MacOS;
+
+internal class MacOSNativeElement : Microsoft.UI.Xaml.FrameworkElement
+{
+	public nint NativeHandle { get; internal set; }
+
+	internal bool Detached { get; set; }
+}
+
+internal class MacOSNativeElementHostingExtension : ContentPresenter.INativeElementHostingExtension
+{
+	private readonly ContentPresenter _presenter;
+	private readonly MacOSWindowNative? _window;
+
+	private MacOSNativeElementHostingExtension(ContentPresenter contentPresenter)
+	{
+		_presenter = contentPresenter;
+		_window = _presenter.XamlRoot?.HostWindow?.NativeWindow as MacOSWindowNative;
+	}
+
+	public static void Register() => ApiExtensibility.Register<ContentPresenter>(typeof(ContentPresenter.INativeElementHostingExtension), o => new MacOSNativeElementHostingExtension(o));
+
+	public void ArrangeNativeElement(object content, Rect arrangeRect, Rect clipRect)
+	{
+		if (content is MacOSNativeElement element)
+		{
+			if (element.Detached)
+			{
+				this.Log().Debug($"Cannot arrange element `{nameof(content)}` of type {content.GetType().FullName} since it was detached.");
```

Comment:
Maybe also a `Debug.Assert`?

Reply:
It's not a theoretical issue as it happens while running SampleApp. Using `Debug.Assert` would be harmful for normal (app) debugging.